crates/engine/src/io.rs

```diff
@@ -1,22 +1,215 @@
-use std::sync::{Arc, RwLock};
-use wasi_common::pipe::{ReadPipe, WritePipe};
-
-/// Input/Output stream redirects
-#[derive(Clone)]
-pub struct IoStreamRedirects {
-    /// Standard input redirect.
-    pub stdin: ReadPipe<std::io::Cursor<Vec<u8>>>,
-    /// Standard output redirect.
-    pub stdout: OutRedirect,
-    /// Standard error redirect.
-    pub stderr: OutRedirect,
-}
-
-/// Output redirect and lock.
-#[derive(Clone)]
-pub struct OutRedirect {
-    /// Output redirect.
-    pub out: WritePipe<Vec<u8>>,
-    /// Lock for writing.
-    pub lock: Arc<RwLock<Vec<u8>>>,
+use std::{
+    io::{LineWriter, Write},
+    sync::{Arc, RwLock, RwLockReadGuard},
+};
+use wasi_common::{
+    pipe::{ReadPipe, WritePipe},
+    WasiFile,
+};
+
+/// Which components should have their logs followed on stdout/stderr.
+#[derive(Clone, Debug)]
+pub enum FollowComponents {
+    /// None of them.
+    None,
+    /// Only those listed.
+    Named(Vec<String>),
+    /// All of them.
+    All,
+}
+
+impl FollowComponents {
+    /// Whether a given component should have its logs followed on stdout/stderr.
+    #[allow(clippy::ptr_arg)] // Clippy recommendation makes it not compile
```

Comment:
The [docs suggest](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.contains) using `any`:

```rust
ids.any(|id| id == component_id)
```